base/sort.jl

```diff
@@ -1285,19 +1285,101 @@ function sort!(A::AbstractArray;
     nd = ndims(A)
-    k = dims
 
-    1 <= k <= nd || throw(ArgumentError("dimension out of range"))
+    1 <= K <= nd || throw(ArgumentError("dimension out of range"))
 
-    remdims = ntuple(i -> i == k ? 1 : axes(A, i), nd)
+    remdims = ntuple(i -> i == K ? 1 : axes(A, i), nd)
     for idx in CartesianIndices(remdims)
-        Av = view(A, ntuple(i -> i == k ? Colon() : idx[i], nd)...)
-        sort!(Av, alg, ordr)
+        Av = view(A, ntuple(i -> i == K ? Colon() : idx[i], nd)...)
+        sort!(Av, alg, order)
     end
     A
 end
+"""
+    sortperm(A; dims::Integer, alg::Algorithm=DEFAULT_UNSTABLE, lt=isless, by=identity, rev::Bool=false, order::Ordering=Forward)
+
+Return a permutation vector `I` that puts `v[I]` in sorted order along the given dimension. The order is specified
+using the same keywords as [`sort!`](@ref). The permutation is guaranteed to be stable even
+if the sorting algorithm is unstable, meaning that indices of equal elements appear in
+ascending order.
+
+To sort slices of an array, refer to [`sortslices`](@ref).
+
+# Examples
+```jldoctest
+julia> A = [8 7; 5 6]
+2×2 Matrix{Int64}:
+ 8  7
+ 5  6
+
+julia> sortperm(A, dims = 1)
+2×2 Matrix{Int64}:
+ 2  4
+ 1  3
+
+julia> sortperm(A, dims = 2)
+2×2 Matrix{Int64}:
+ 3  1
+ 2  4
+```
+"""
+function sortperm(A::AbstractArray; dims::Integer,
+                  alg::Algorithm=DEFAULT_UNSTABLE,
+                  lt=isless,
+                  by=identity,
+                  rev::Union{Bool,Nothing}=nothing,
+                  order::Ordering=Forward)
+    ix = Base.copymutable(LinearIndices(A))
+    sortperm!(ix,A; dims, alg, lt, by, rev, order, initialized = true)
+end
+
+"""
+    sortperm!(ix, A; dims::Integer, alg::Algorithm=DEFAULT_UNSTABLE, lt=isless, by=identity, rev::Bool=false, order::Ordering=Forward, initialized::Bool=false)
+
+Like [`sortperm`](@ref), but accepts a preallocated index Array `ix`.  If `initialized` is `false`
+(the default), `ix` is initialized to contain the values `LinearIndices(A)`.
+
+# Examples
+```jldoctest
+julia> A = [8 7; 5 6]; p = zeros(Int,2, 2);
+
+julia> sortperm!(p, A;dims=1); p
+2×2 Matrix{Int64}:
+ 2  4
+ 1  3
+
+julia> sortperm!(p, A;dims=2); p
+2×2 Matrix{Int64}:
+ 3  1
+ 2  4
+```
+"""
+function sortperm!(ix::AbstractArray{<:Integer}, A::AbstractArray;
+                    dims::Integer,
+                    alg::Algorithm=DEFAULT_UNSTABLE,
+                    lt=isless,
+                    by=identity,
+                    rev::Union{Bool,Nothing}=nothing,
+                    order::Ordering=Forward,
+                    initialized::Bool=false)
+    ndims(ix) == ndims(A) || throw(ArgumentError("dimensions of index and source array must be equal")) 
+    for (axes_A, axes_ix) in zip(axes(A),axes(ix))
+        axes_A == axes_ix || throw(ArgumentError("index array must have the same length/indices as the source array, $(axes_A) != $(axes_ix)"))
```

Comment:
```suggestion
    axes(ix) == axes(A) || throw(ArgumentError("index array must have the same size/axes as the source array, $(axes(ix)) != $(axes(A))"))
```
Tuple equality does both these checks internally